File: backend/app/api/schemas/document.py
"""Pydantic schemas for the Document API."""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict

from app.domain.enums import DocumentStatus


class DocumentUploadResponse(BaseModel):
    document_id: UUID
    status: DocumentStatus
    page_count: int


class DocumentResponse(BaseModel):
    """Full document representation returned by GET /{document_id}."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    knowledge_base_id: UUID
    filename: str
    content_type: str
    byte_size: int
    page_count: int | None = None
    status: DocumentStatus
    title: str | None = None
    checksum: str | None = None
    language: str
    failure_reason: str | None = None
    created_at: datetime
    updated_at: datetime
    processed_at: datetime | None = None


class DocumentStatusResponse(BaseModel):
    """Processing-state snapshot returned by GET /{document_id}/status."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    status: DocumentStatus
    page_count: int | None = None
    failure_reason: str | None = None
    updated_at: datetime


class DocumentUrlResponse(BaseModel):
    """Time-limited download URL returned by GET /{document_id}/url."""

    url: str
    expires_at: datetime


class BoundingBoxResponse(BaseModel):
    x0: float
    y0: float
    x1: float
    y1: float


class DocumentRegionResponse(BaseModel):
    """A table or figure bounding box the student can click to focus a question."""

    id: UUID
    region_type: str  # "table" or "figure"
    page_number: int
    bounding_box: BoundingBoxResponse

File: frontend/src/schemas/document.ts
import { z } from 'zod';

import { documentStatus } from '@/schemas/enums';
import { instant, uuid } from '@/schemas/primitives';

/** Document responses. Optional fields are nullable for the reason given alongside the
 * Knowledge Base schemas: the backend sends an explicit null rather than omitting a key. */

/** What an accepted upload reports back, before any of the file has been read. */
export const documentUpload = z.object({
  document_id: uuid,
  status: documentStatus,
  page_count: z.number().int(),
});
export type DocumentUpload = z.infer<typeof documentUpload>;

export const document = z.object({
  id: uuid,
  knowledge_base_id: uuid,
  filename: z.string(),
  content_type: z.string(),
  byte_size: z.number().int(),
  page_count: z.number().int().nullable(),
  status: documentStatus,
  title: z.string().nullable(),
  checksum: z.string().nullable(),
  language: z.string(),
  failure_reason: z.string().nullable(),
  created_at: instant,
  updated_at: instant,
  processed_at: instant.nullable(),
});
export type Document = z.infer<typeof document>;

export const documentList = z.array(document);

/** Time-limited presigned URL returned by GET /{document_id}/url. */
export const documentUrl = z.object({
  url: z.string(),
  expires_at: instant,
});
export type DocumentUrl = z.infer<typeof documentUrl>;

/** A table or figure bounding box the student can click to focus a question. */
export const documentRegion = z.object({
  id: uuid,
  region_type: z.enum(['table', 'figure']),
  page_number: z.number().int(),
  bounding_box: z.object({
    x0: z.number(),
    y0: z.number(),
    x1: z.number(),
    y1: z.number(),
  }),
});
export type DocumentRegion = z.infer<typeof documentRegion>;

export const documentRegionList = z.array(documentRegion);

/**
 * The smaller shape the status endpoint returns, which is the one a screen polls while a
 * document is being read. It carries the failure reason because a poll that only reported
 * FAILED would leave the reader to guess, and guessing at someone else's file is unkind.
 */
export const documentStatusSnapshot = z.object({
  id: uuid,
  status: documentStatus,
  page_count: z.number().int().nullable(),
  failure_reason: z.string().nullable(),
  updated_at: instant,
});
export type DocumentStatusSnapshot = z.infer<typeof documentStatusSnapshot>;
